src/tribler-core/tribler_core/modules/trust_calculation/trust_graph.py

```diff
@@ -0,0 +1,141 @@
+            raise TrustGraphException("Max node peers reached in graph")
+
+        # Node does not exist in the graph so a new node at this point.
+        # The numeric node id is used here so the id for the new node becomes
+        # equal to the number of nodes in the graph.
+        node_id = self.number_of_nodes()
+        node_attrs = {
+            'id': node_id,
+            'key': hexlify(peer_key),
+            'total_up': self.bandwidth_db.get_total_given(peer_key),
+            'total_down': self.bandwidth_db.get_total_taken(peer_key)
+        }
+        self.add_node(node_id, **node_attrs)
+        self.node_public_keys.append(peer_key)
+
+        return self.nodes()[node_id]
+
+    def compose_graph_data(self):
+        layer_1 = self.bandwidth_db.get_latest_transactions(self.root_key)
+        for tx in layer_1:
+            self.add_bandwidth_transaction(tx)
+
+            # Stop at layer 2
+            counter_party = tx.public_key_a if self.root_key != tx.public_key_a else tx.public_key_b
+            layer_2 = self.bandwidth_db.get_latest_transactions(counter_party)
+            for tx2 in layer_2:
+                self.add_bandwidth_transaction(tx2)
+
+    def compute_edge_id(self, transaction):
+        sha2 = hashlib.sha3_224()  # any safe hashing should do
+        sha2.update(transaction.public_key_a)
+        sha2.update(transaction.public_key_b)
+        return sha2.hexdigest()[:64]
+
+    def add_bandwidth_transaction(self, tx):
+        # First, compose a unique edge id for the transaction and check if it is already added.
+        edge_id = self.compute_edge_id(tx)
+
+        if edge_id not in self.edge_set:
+            peer1 = self.get_or_create_node(tx.public_key_a, add_if_not_exist=True)
+            peer2 = self.get_or_create_node(tx.public_key_b, add_if_not_exist=True)
+
+            if peer2['id'] not in self.successors(peer1['id']):
+                self.add_edge(peer1['id'], peer2['id'])
+
+            self.edge_set.add(edge_id)
+
+    def compute_node_graph(self):
+        undirected_graph = self.to_undirected()
+        num_nodes = undirected_graph.number_of_nodes()
+
+        # Find bfs tree of the connected components
+        bfs_tree = nx.bfs_tree(undirected_graph, ROOT_NODE_ID)
+
+        # Position the nodes in a circular fashion according to the bfs tree
+        pos = GraphPositioning.hierarchy_pos(bfs_tree, root=ROOT_NODE_ID, width=2 * math.pi, xcenter=0.5)
+
+        graph_nodes = []
+        graph_edges = []
+        index_mapper = {}
+
+        node_id = ROOT_NODE_ID
+        max_x = max_y = 0.0001  # as close to zero
+        for _id, (theta, r) in pos.items():
+            index_mapper[_id] = node_id
+            node = undirected_graph.nodes()[_id]
+            node['id'] = node_id
+            node_id += 1
+
+            # convert from polar coordinates to cartesian coordinates
+            x = r * math.sin(theta) * num_nodes
+            y = r * math.cos(theta) * num_nodes
```

Comment:
wow! you do it manually...

Reply:
Credit goes to @cuileri for the first implementation.